File: public/java/099-time-based-key-value-store.java
/*
 * LeetCode Problem #981: Time Based Key-Value Store
 * Link: https://leetcode.com/problems/time-based-key-value-store/
 * Difficulty: Medium
 */
import java.util.*;

class Solution {

    // APPROACH 1: HashMap + Binary Search | O(1) set | O(log n) get | O(n) space
    // EXPLAIN: Store (timestamp, value) pairs per key in insertion order; binary search for the largest ts ≤ query.
    static class TimeMap1 {
        Map<String, List<int[]>> store = new HashMap<>();      // key -> [(timestamp, valueIdx)]
        Map<String, List<String>> vals = new HashMap<>();

        public void set(String key, String value, int timestamp) {
            store.computeIfAbsent(key, k -> new ArrayList<>()).add(new int[]{timestamp});
            vals.computeIfAbsent(key, k -> new ArrayList<>()).add(value);
        }

        public String get(String key, int timestamp) {
            if (!store.containsKey(key)) return "";
            List<int[]> ts = store.get(key);
            int lo = 0, hi = ts.size() - 1;
            String result = "";
            while (lo <= hi) {
                int mid = lo + (hi - lo) / 2;
                if (ts.get(mid)[0] <= timestamp) { result = vals.get(key).get(mid); lo = mid + 1; }
                else hi = mid - 1;
            }
            return result;
        }
    }

    // APPROACH 2: HashMap + TreeMap | O(log n) set | O(log n) get | O(n) space
    // EXPLAIN: Use a TreeMap per key so floorKey(timestamp) directly returns the largest valid timestamp.
    static class TimeMap2 {
        Map<String, TreeMap<Integer, String>> store = new HashMap<>();

        public void set(String key, String value, int timestamp) {
            store.computeIfAbsent(key, k -> new TreeMap<>()).put(timestamp, value);
        }

        public String get(String key, int timestamp) {
            if (!store.containsKey(key)) return "";
            Map.Entry<Integer, String> entry = store.get(key).floorEntry(timestamp);
            return entry == null ? "" : entry.getValue();
        }
    }

    // APPROACH 3: HashMap + Linear Search | O(1) set | O(n) get | O(n) space
    // EXPLAIN: Append to a list per key; scan from the end to find the latest valid timestamp.
    static class TimeMap3 {
        Map<String, List<int[]>> times = new HashMap<>();
        Map<String, List<String>> vals = new HashMap<>();

        public void set(String key, String value, int timestamp) {
            times.computeIfAbsent(key, k -> new ArrayList<>()).add(new int[]{timestamp});
            vals.computeIfAbsent(key, k -> new ArrayList<>()).add(value);
        }

        public String get(String key, int timestamp) {
            if (!times.containsKey(key)) return "";
            List<int[]> ts = times.get(key);
            String result = "";
            for (int i = 0; i < ts.size(); i++) {
                if (ts.get(i)[0] <= timestamp) result = vals.get(key).get(i);
                else break;
            }
            return result;
        }
    }

    // APPROACH 4: HashMap + Parallel Lists (upper_bound via manual binary search) | O(1) set | O(log n) get | O(n) space
    // EXPLAIN: Maintain parallel timestamp and value lists per key; binary search for the insertion point, step back one.
    static class TimeMap4 {
        Map<String, List<Integer>> times = new HashMap<>();
        Map<String, List<String>> vals = new HashMap<>();

        public void set(String key, String value, int timestamp) {
            times.computeIfAbsent(key, k -> new ArrayList<>()).add(timestamp);
            vals.computeIfAbsent(key, k -> new ArrayList<>()).add(value);
        }

        public String get(String key, int timestamp) {
            if (!times.containsKey(key)) return "";
            List<Integer> ts = times.get(key);
            // bisect_right: find first index where ts[i] > timestamp
            int lo = 0, hi = ts.size();
            while (lo < hi) {
                int mid = lo + (hi - lo) / 2;
                if (ts.get(mid) <= timestamp) lo = mid + 1;
                else hi = mid;
            }
            int idx = lo - 1;
            return idx < 0 ? "" : vals.get(key).get(idx);
        }
    }

    // APPROACH 5: HashMap + HashMap (Nested) with Timestamp Scan | O(1) set | O(n) get | O(n) space
    // EXPLAIN: Store key -> {timestamp -> value} and a separate list of timestamps; scan for best match.
    static class TimeMap5 {
        Map<String, Map<Integer, String>> store = new HashMap<>();
        Map<String, List<Integer>> timestamps = new HashMap<>();

        public void set(String key, String value, int timestamp) {
            store.computeIfAbsent(key, k -> new HashMap<>()).put(timestamp, value);
            timestamps.computeIfAbsent(key, k -> new ArrayList<>()).add(timestamp);
        }

        public String get(String key, int timestamp) {
            if (!store.containsKey(key)) return "";
            int bestTime = -1;
            for (int t : timestamps.get(key))
                if (t <= timestamp && t > bestTime) bestTime = t;
            return bestTime == -1 ? "" : store.get(key).get(bestTime);
        }
    }
}

// Made with Bob

File: public/python/099-time-based-key-value-store.py
# LeetCode Problem #981: Time Based Key-Value Store
# Difficulty: Medium
# Link: https://leetcode.com/problems/time-based-key-value-store/

from typing import List
import bisect
from collections import defaultdict

# ─────────────────────────────────────────────
# APPROACH 1: Hash Map + Binary Search | O(1) set | O(log n) get | O(n) space
# EXPLAIN: Store (timestamp, value) pairs per key in insertion order; binary search for the largest timestamp ≤ query.
# WHEN: Optimal — set is O(1) amortised, get is O(log n); standard solution.

class TimeMap1:
    def __init__(self):
        self.store: dict = defaultdict(list)  # key -> [(timestamp, value)]

    def set(self, key: str, value: str, timestamp: int) -> None:
        self.store[key].append((timestamp, value))

    def get(self, key: str, timestamp: int) -> str:
        if key not in self.store:
            return ""
        entries = self.store[key]
        lo, hi = 0, len(entries) - 1
        result = ""
        while lo <= hi:
            mid = lo + (hi - lo) // 2
            if entries[mid][0] <= timestamp:
                result = entries[mid][1]
                lo = mid + 1
            else:
                hi = mid - 1
        return result

# ─────────────────────────────────────────────
# APPROACH 2: Hash Map + Sorted Map (dict keys sorted) | O(log n) set | O(log n) get | O(n) space
# EXPLAIN: Use a plain dict per key and maintain a sorted list of timestamps for bisect lookup.
# WHEN: When you want explicit timestamp-key sorted access without a tree map library.

class TimeMap2:
    def __init__(self):
        self.store: dict = defaultdict(dict)    # key -> {timestamp: value}
        self.ts_list: dict = defaultdict(list)  # key -> sorted [timestamps]

    def set(self, key: str, value: str, timestamp: int) -> None:
        self.store[key][timestamp] = value
        bisect.insort(self.ts_list[key], timestamp)

    def get(self, key: str, timestamp: int) -> str:
        if key not in self.store:
            return ""
        ts = self.ts_list[key]
        idx = bisect.bisect_right(ts, timestamp) - 1
        if idx < 0:
            return ""
        return self.store[key][ts[idx]]

# ─────────────────────────────────────────────
# APPROACH 3: Hash Map + Linear Search | O(1) set | O(n) get | O(n) space
# EXPLAIN: Append to a list per key; scan from the end to find the latest valid timestamp.
# WHEN: Simplest implementation; only feasible when get calls are rare or n is very small.

class TimeMap3:
    def __init__(self):
        self.store: dict = defaultdict(list)

    def set(self, key: str, value: str, timestamp: int) -> None:
        self.store[key].append((timestamp, value))

    def get(self, key: str, timestamp: int) -> str:
        result = ""
        for ts, val in self.store.get(key, []):
            if ts <= timestamp:
                result = val
            else:
                break
        return result

# ─────────────────────────────────────────────
# APPROACH 4: Hash Map + bisect upper_bound | O(1) set | O(log n) get | O(n) space
# EXPLAIN: Use bisect.bisect_right on a parallel timestamp list to locate the insertion point, then step back one.
# WHEN: Equivalent to Approach 1 but uses bisect library directly for cleaner code.

class TimeMap4:
    def __init__(self):
        self.vals: dict = defaultdict(list)  # key -> [value]
        self.times: dict = defaultdict(list) # key -> [timestamp]

    def set(self, key: str, value: str, timestamp: int) -> None:
        self.times[key].append(timestamp)
        self.vals[key].append(value)

    def get(self, key: str, timestamp: int) -> str:
        if key not in self.times:
            return ""
        idx = bisect.bisect_right(self.times[key], timestamp) - 1
        if idx < 0:
            return ""
        return self.vals[key][idx]

# ─────────────────────────────────────────────
# APPROACH 5: Nested Hash Maps with Timestamp Scan | O(1) set | O(n) get | O(n) space
# EXPLAIN: Store key -> {timestamp -> value} and separately track all timestamps; scan for best match.
# WHEN: When direct timestamp lookup by key is needed and get performance is not critical.

class TimeMap5:
    def __init__(self):
        self.store: dict = defaultdict(dict)
        self.timestamps: dict = defaultdict(list)

    def set(self, key: str, value: str, timestamp: int) -> None:
        self.store[key][timestamp] = value
        self.timestamps[key].append(timestamp)

    def get(self, key: str, timestamp: int) -> str:
        if key not in self.store:
            return ""
        best_time = -1
        for t in self.timestamps[key]:
            if t <= timestamp and t > best_time:
                best_time = t
        return "" if best_time == -1 else self.store[key][best_time]

# Made with Bob
